ibis/util.py

```diff
@@ -217,7 +217,11 @@ def is_iterable(o: Any) -> bool:
     >>> is_iterable([])
     True
     """
-    return not isinstance(o, (str, bytes)) and isinstance(o, collections.abc.Iterable)
+    from ibis.expr.deferred import Deferred
```

Comment:
Could we check for deferreds before calling `is_iterable()` in the call site instead? I assume it is checked in `bind_expr()`?

Reply:
It's used in a bunch of places. It's a lot more fragile than to check it in the core method that's powering all the call sites.

Reply:
What's the issue with having it here? The import?

Reply:
I have two alternative implementation suggestions:
1. investigate the call sites and first check for deferreds
2. have a `NotIterable` ABC defined here, and extend that from `Deferred`, we could also add `NotIterable.register(str|bytes)`

Reply:
I don't think it indicates a design issue at the call site as much as a limitation of using duck typing to define an interface: you can't rely on method **existence** to accurately determine whether a thing implements a protocol.

Reply:
In any case, it seems strictly worse to introduce a bunch of call-site checks that are likely to be forgotten.

Reply:
I agree, though we could also follow the `try first` philosophy instead of checking with `is_iterable()`, that way deferred would clearly raise.

Reply:
> In any case, it seems strictly worse to introduce a bunch of call-site checks that are likely to be forgotten.

How about option 2?

Reply:
Seems fine! Lemme implement it.